core/src/main/java/jenkins/tasks/SimpleBuildStep.java

```diff
@@ -77,15 +78,15 @@
     /**
      * Run this step.
      * @param run a build this is running as a part of
-     * @param workspace a workspace to use for any file operations
-     * @param launcher a way to start processes
+     * @param workspace a workspace to use for any file operations; may be {@code null}, unless {@link #requiresWorkspace()} returns {@code true}
+     * @param launcher a way to start processes; may be {@code null}, unless {@link #requiresLauncher()} returns {@code true}
      * @param listener a place to send output
      * @throws InterruptedException if the step is interrupted
      * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
      * @deprecated Use {@link #perform(Run, FilePath, EnvVars, Launcher, TaskListener) instead.}
      */
     @Deprecated
-    default void perform(@NonNull Run<?, ?> run, @NonNull FilePath workspace, @NonNull Launcher launcher,
+    default void perform(@NonNull Run<?, ?> run, @CheckForNull FilePath workspace, @CheckForNull Launcher launcher,
```

Comment:
```suggestion
    default void perform(@NonNull Run<?, ?> run, @Nullable FilePath workspace, @Nullable Launcher launcher,
```

since whether or not null is actually allowed is dependent on complex things the static analyzer will not understand

Reply:
It was not particularly clear to me when each of those two was applicable; I rather assumed they were equivalent so I just used the same one everywhere. Will have a look around and adjust these.

Reply:
> The annotated element could be null under some circumstances.
> In general, this means developers will have to read the documentation to determine when a null value is acceptable and whether it is necessary to check for a null value.

Reply:
Switched `@CheckForNull` to `@Nullable` as suggested